networkd: disallow useDHCP

globin opened the pull request:
###### Motivation for this change
This is preparation to switch to networkd by default in 20.03
Also intended for 19.09, does not change anything in the default config.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @fpletz @flokli @arianvp @disassembler @lheckemann

globin committed:
networking.useDHCP: disallow for networkd

This setting will be removed with the switch to systemd-networkd. The
use of per interface config is encouraged instead.

volth commented:
`initrd-network` depends on `networking.useDHCP`
https://github.com/NixOS/nixpkgs/blob/2c3dcbb9d09a9757c26a1bc34ba5b8f263a7e554/nixos/modules/system/boot/initrd-network.nix

And... will initrd switch to `networkd` too?
Otherwise I'd prefer to stay with scripted networking only for the reason to have the same settings for initrd and non-initrd networking

fpletz commented:
@globin @andir Regarding https://github.com/NixOS/nixpkgs/pull/68096/commits/2b605e96c289b5f43fc31f3c177d6622a2767142, with this change we also can remove the `99-main.network` hack altogether.

As mentioned elsewhere this functionality is inherently broken without manually poking around some other network units anyway; for example to make stuff like `network-online.target` work if you have more than one interface on your system (e.g. `RequiredForOnline=`).

globin committed:
networking.useDHCP: add release notes and docs

globin commented:
Added release notes and information to the option docs.

andir commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -903,6 +903,11 @@ in
         configuration for all network interfaces that are not manually
         configured.
+
+        Using this option is highly discouraged and also incompatible with
```
Might be a bit far fetched but how do we plan to proceed with installer images if we discourage usage of this? Wouldn't it be best practice to ship the installer with an encouraged configuration? We should be a good example to users and not just tell them to not use it.

globin replied:
Currently working on this

arianvp replied:
Does `networking.interfaces."en*".useDHCP = true` work? reason I ask is because setting a wildcard for `matchConfig` is supported for networkd. 

I guess then for the installer image we can then set `networking.interfaces."en*".useDHCP = true` to get an IP address on all wired connections (seems like a sane default), but it won't try to ask for DHCP leases on bridges, veths etc

fpletz replied:
IMHO we shouldn't solve this in this PR.

globin replied:
@andir I understood your comment as changing nixos-generate-config, from the conversation I just had he misunderstood this, so I've added a commit to change nixos-generate-config. It adds `networking.useDHCP = false;` and adds `networking.interfaces.<name>.useDHCP = true;` for all interfaces is `/sys/class/net` except for `lo`. Virtual interfaces should not be present in our install media anyway.

globin committed:
nixos-generate-config: add useDHCP per interface

This sets networking.useDHCP to false and for all interfaces found the
per-interface useDHCP to true. This replicates the current default
behaviour and prepares for the switch to networkd.

flokli commented on nixos/tests/networking.nix:
```diff
@@ -72,6 +72,7 @@ let
   testCases = {
     loopback = {
       name = "Loopback";
+      machine.networking.useDHCP = false;
```
can't we just stip all lines setting `networking.useDHCP = false;` in the tests, if we set it for each interface?

globin replied:
We now assert `useDHCP` to be false in conjunction with networkd, I didn't want to change the whole test suite that for the most part doesn't use networkd.

globin committed:
Merge branch 'master' into networkd-disallow-dhcp

lheckemann commented:
@GrahamcOfBorg test networking ipv6

lheckemann commented:
Backported: 907bb84e4b3 da9e914b6c0 1cb925e8a1f

lheckemann commented:
Looks like I was a bit hasty with merging this — it broke evaluation for the networkd variant of the `networking.virtual` test.

arianvp commented:
@lheckemann does that mean you reverted those changes for `19.09` or are they still in?

globin commented:
They are included. The test never ran with networkd, I simply added the `useNetworkd = networkd` and the failing test probably is a valid issue with our networkd config.

lheckemann commented:
@arianvp I fixed the evaluation but not the test, 4a03ddd08d47dccac1158d37272e137c53e8b35b

matthewbauer commented:
I think this makes hotplugging network devices difficult. The interface name is different depending on which port you have. It takes a long time to boot up if it is missing too (1 min 30 sec by default).

Ma27 commented:
Not sure if I misunderstand you, but can't you use wildcards for that purpose in `networkd`? With scripted networking, this is still supported IIRC.

lheckemann commented:
Concretely for the networkd case:
```nix
{
    systemd.network.networks."99-en-dhcp" = {
      matchConfig.Name = "en*";
      networkConfig.DHCP = "yes";
    };
}
```

to get DHCP on all ethernet devices.

matthewbauer commented:
@lheckemann Thanks! That helps.

Does nixos-generate-config (https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/tools/nixos-generate-config.pl#L570-L586) need to hardcode these interfaces at all then? If one of those interfaces no longer exists, you get stuck waiting for "network-addresses-en*" or "network-addresses-wl*" to finish (or timeout at 1m 30s) as part of network-setup.service. This could happen frequently if you're switching between ethernet and wifi but have a USB adapter.

arianvp commented:
A wildcard sounds fine to me instead of a fixed interface

lheckemann commented:
The problem with that approach is that it doesn't work with scripted networking, which is still our default.

matthewbauer commented:
Opened https://github.com/NixOS/nixpkgs/issues/107908 to track this.
